pkg/extensions/cluster_test.go

```diff
@@ -62,15 +72,304 @@ var _ = Describe("Cluster", func() {
 		})
 	})
 
-	Describe("#GetShootStateForCluster", func() {
-		var (
-			expectedShoot      *gardencorev1beta1.Shoot
-			expectedShootState *gardencorev1beta1.ShootState
+	Describe("#SyncClusterResourceToSeed", func() {
+		BeforeEach(func() {
+			expectedCloudProfile = &gardencorev1beta1.CloudProfile{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: "core.gardener.cloud/v1beta1",
+					Kind:       "CloudProfile",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "foo",
+				},
+			}
+
+			expectedSeed = &gardencorev1beta1.Seed{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: "core.gardener.cloud/v1beta1",
+					Kind:       "Seed",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "foo",
+				},
+			}
+
+			expectedShoot = &gardencorev1beta1.Shoot{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: "core.gardener.cloud/v1beta1",
+					Kind:       "Shoot",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "foo",
+					Namespace: "garden-bar",
+				},
+			}
+
+			clusterName = "shoot--" + expectedShoot.Namespace + "--" + expectedShoot.Name
+			cluster = &extensionsv1alpha1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: clusterName,
+				},
+			}
+		})
+
+		It("should not sync cloudprofile, shoot and seed to cluster if seed is not assigned to shoot", func() {
+			Expect(fakeGardenClient.Create(ctx, expectedCloudProfile)).To(Succeed())
+			Expect(fakeGardenClient.Create(ctx, expectedSeed)).To(Succeed())
+			Expect(fakeGardenClient.Create(ctx, expectedShoot)).To(Succeed())
+			Expect(fakeSeedClient.Create(ctx, cluster)).To(Succeed())
 
-			clusterName string
-			cluster     *extensionsv1alpha1.Cluster
-		)
+			Expect(SyncClusterResourceToSeed(ctx, fakeSeedClient, cluster.Name, expectedShoot, expectedCloudProfile, expectedSeed)).To(Succeed())
+			Expect(fakeSeedClient.Get(ctx, client.ObjectKeyFromObject(cluster), cluster)).To(Succeed())
 
+			Expect(cluster.Spec.CloudProfile.Object).To(BeNil())
+			Expect(cluster.Spec.Seed.Object).To(BeNil())
+			Expect(cluster.Spec.Shoot.Object).To(BeNil())
+		})
+
+		It("should sync cloudprofile, shoot and seed to cluster", func() {
+			expectedShoot.Spec.SeedName = pointer.String(expectedSeed.Name)
+
+			Expect(fakeGardenClient.Create(ctx, expectedCloudProfile)).To(Succeed())
+			Expect(fakeGardenClient.Create(ctx, expectedSeed)).To(Succeed())
+			Expect(fakeGardenClient.Create(ctx, expectedShoot)).To(Succeed())
+			Expect(fakeSeedClient.Create(ctx, cluster)).To(Succeed())
```

Comment:
Same as above, shouldn't we also check if the cluster gets created now if it didn't exist before?